samples/algorithms/variational-algorithms/Optimization.qs

```diff
@@ -0,0 +1,164 @@
+    ///
+    /// # Named Items
+    /// ## StepScale
+    /// The coefficient by which steps along gradient vectors should be scaled.
+    /// ## StepPower
+    /// The power to which the iteration number should be raised when computing
+    /// how far to step along the gradient vector.
+    /// ## StepOffset
+    /// A number to be added to the number of iterations when computing
+    /// how far to step along the gradient vector.
+    /// ## SearchScale
+    /// The coefficient by which searches should be scaled when estimating
+    /// gradient vectors.
+    /// ## SearchPower
+    /// The power to which the iteration number should be raised when computing
+    /// how far to search in order to estimate gradient vectors.
+    /// ## NIterations
+    /// The number of iterations of SPSA to run before stopping.
+    /// ## MaximumSetback
+    /// Whether the maximum setback rule is enabled (requiring an additional
+    /// objective evaluation at each iteration), and if so, the maximum
+    /// allowed increase in objective values at each iteration.
+    /// ## VerboseMessage
+    /// A function to be called to report verbose information about optimization
+    /// progress.
+    newtype SpsaOptions = (
+         StepScale: Double,
+         StepPower: Double,
+         StepOffset: Int,
+         SearchScale: Double,
+         SearchPower: Double,
+         NIterations: Int,
+         MaximumSetback: (Bool, Double),
+         VerboseMessage: String -> Unit
+    );
+
+    /// # Summary
+    /// Returns a default set of options for use with SPSA optimization.
+    function DefaultSpsaOptions() : SpsaOptions {
+        return Default<SpsaOptions>()
+            w/ SearchScale <- 0.1
+            w/ SearchPower <- 0.101
+            w/ StepScale <- 1.0
+            w/ StepPower <- 0.602
+            w/ StepOffset <- 0
+            w/ MaximumSetback <- (false, 0.1)
+            w/ NIterations <- 30
+            w/ VerboseMessage <- Message;
+    }
+
+    /// # Summary
+    /// Given an operation that evaluates an objective at a given point,
+    /// attempts to find the minimum value of the objective by using the
+    /// simulntaneous perturbative stochastic approximation (SPSA).
+    ///
+    /// # Input
+    /// ## oracle
+    /// An operation that evaluates the objective function at a given point.
+    /// ## startingPoint
+    /// An initial guess to be used in optimizing the objective function
+    /// provided.
+    /// ## options
+    /// Options used to control the optimization algorithm.
+    ///
+    /// # Output
+    /// The coordinates and final objective value found by the SPSA algorithm.
+    operation FindMinimumWithSpsa(oracle : (Double[] => Double), startingPoint : Double[], options : SpsaOptions) : (Double[], Double) {
+        let nParameters = Length(startingPoint);
+        let deltaDist = DiscreteUniformDistribution(0, 1);
+        let drawDelta = Delayed(MaybeChooseElement, ([-1.0, 1.0], deltaDist));
```

Comment:
It would be helpful to explain this line with a comment.

Reply:
Good point, let me go add that then. Thanks!